Mark legacy exporter features in spec compliance optional. SDKs are free to su… (#1359)
open-telemetry/opentelemetry-specification: spec-compliance-matrix.md

## Traces

> |Feature                                       |Go |Java|JS |Python|Ruby|Erlang|PHP|Rust|C++|.Net|Swift|Optional|

Reviewer: `Optional` column definitely should be the second column, right after `Feature`, not the last one :)

|End with timestamp                            | + | +  | + | +    | +  | +    | + | +  | + | +  | +   |        |
|IsRecording                                   | + | +  | + | +    | +  | +    | + | +  | + | +  | +   |        |
|IsRecording becomes false after End           |   | +  | + | +    | +  | +    |   |    |   |    | +   |        |
|Set status with StatusCode (Unset, Ok, Error) |   | +  | + | +    | +  | -    |   | +  |   | +  | +   |        |
|Safe for concurrent calls                     | + | +  | + | +    | +  | +    | + | +  | + | +  | +   |        |
|events collection size limit                  |   | +  | + | +    | +  | -    |   | +  |   | -  | +   |        |
|attribute collection size limit               |   | +  | + | +    | +  | -    |   | +  |   | -  | +   |        |
|links collection size limit                   |   | +  | + | +    | +  | -    |   | +  |   | -  | +   |        |
|[Span attributes](specification/trace/api.md#set-attributes)|
|SetAttribute                                  | + | +  | + | +    | +  | +    | + | +  | + | +  | +   |        |

> |Set order preserved                           | + | -  | + | +    | +  | +    | + | +  | + | +  | +   |        |

Reviewer 2: I believe this is optional in the spec.

> | Set order preserved                                                                              | X        | +  | -    | +  | +      | +    | +      | +   | +    | +   | +    | +     |

Reviewer 3: Should we use `YES` instead of `X` for the optional values?

Author: Have no preference for either, so for now sticking with what's here unless someone has stronger preference :)

Reviewer 4: I'd prefer just writing _Optional_ in there (the column has the width of the header anyway), so one would not have to look up the heading. Anyway, I'd rather merge this asap to avoid conflicts and consider prettifying it afterwards in a smaller changeset.

| Double floating-point type                                                                       |          | +  | +    | +  | +      | +    | +      | -   | +    | +   | +    | +     |
| Signed int64 type                                                                                |          | +  | +    | +  | +      | +    | +      | -   | +    | +   | +    | +     |
| Array of primitives (homogeneous)                                                                |          | +  | +    | +  | +      | +    | +      | +   | +    | +   | +    | +     |
| `null` values documented as invalid/undefined                                                    |          |    | +    | +  | +      |      | N/A    |     |      | +   |      | N/A   |
| Unicode support for keys and string values                                                       |          | +  | +    | +  | +      | +    | +      | +   | +    | +   | +    | +     |
| [Span linking](specification/trace/api.md#specifying-links)                                      |          |    |      |    |        |      |        |     |      |     |      |       |

> | AddLink                                                                                          |          | +  | +    | +  | +      | +    | +      | +   | +    | -   | +    | +     |

Reviewer 4: C++: + to - on purpose?

Reviewer 5: Ooops, I just realized we just had a race condition, my bad :( Would you mind filling issues (or PRs) to follow these two details up?

Reviewer 4: No worries, I think it's fine actually, as it seems to be indeed missing:
https://github.com/open-telemetry/opentelemetry-cpp/blob/main/api/include/opentelemetry/trace/span.h
The other change is just cosmetic, nothing urgent.

## Exporters

|Feature                                       |Go |Java|JS |Python|Ruby|Erlang|PHP|Rust|C++|.Net|Swift|Optional|
|----------------------------------------------|---|----|---|------|----|------|---|----|---|----|-----|--------|
|Standard output (logging)                     | + | +  | + | +    | +  | +    | - | +  | + | +  |  +  |        |
|In-memory (mock exporter)                     | + | +  | + | +    | +  | +    | - | -  | - | +  |  +  |        |
|[OTLP](specification/protocol/otlp.md)|
|OTLP/gRPC Exporter                            | + | +  | + | +    |    | +    |   | +  | + | +  |  +  |        |
|OTLP/HTTP binary Protobuf Exporter            | - | -  | + | [-](https://github.com/open-telemetry/opentelemetry-python/issues/1106)    | +  | +    |   |    | + | -  |  -  |   X    |
|OTLP/HTTP JSON Protobuf Exporter              | - | -  | + | [-](https://github.com/open-telemetry/opentelemetry-python/issues/1003)    |    | -    |   |    |   | -  |  -  |   X    |
|OTLP/HTTP gzip Content-Encoding support       | - | -  | + | +    | +  | -    |   |    |   | -  | -   |  X     |

> |Concurrent sending                            | - |    | + | [-](https://github.com/open-telemetry/opentelemetry-python/issues/1108)    |    | -    |   | +  |   | -  | -   |        |

Reviewer 2: is this mandatory?

Author: I figure if an exporter is called twice concurrently there's an expectation that the two sends would happen at the same time in any language.

> |Honors retryable responses with backoff       | + |    | + | +    | +  | -    |   |    |   | -  | -   |        |

Reviewer 2: is this required? If an SDK allows people to configure the grpc channel however they want, does that fulfill this?

Author: Handling retry well seems like a nice feature to support eventually, though not necessarily for 1.0 (which was my intent). I'd think being able to configure the gRPC channel could be enough, and shortcuts could be added based on user demand independently of this spec.

|[OTLP](specification/protocol/otlp.md)|
|OTLP/gRPC Exporter                            | + | +  | + | +    |    | +    |   | +  | + | +  |  +  |        |
|OTLP/HTTP binary Protobuf Exporter            | - | -  | + | [-](https://github.com/open-telemetry/opentelemetry-python/issues/1106)    | +  | +    |   |    | + | -  |  -  |   X    |
|OTLP/HTTP JSON Protobuf Exporter              | - | -  | + | [-](https://github.com/open-telemetry/opentelemetry-python/issues/1003)    |    | -    |   |    |   | -  |  -  |   X    |
|OTLP/HTTP gzip Content-Encoding support       | - | -  | + | +    | +  | -    |   |    |   | -  | -   |  X     |
|Concurrent sending                            | - |    | + | [-](https://github.com/open-telemetry/opentelemetry-python/issues/1108)    |    | -    |   | +  |   | -  | -   |        |
|Honors retryable responses with backoff       | + |    | + | +    | +  | -    |   |    |   | -  | -   |        |
|Honors non-retryable responses                | + |    | - | +    | +  | -    |   |    |   | -  | -   |        |
|Honors throttling response                    | + |    | - | +    | +  | -    |   |    |   | -  | -   |        |

> |Multi-destination spec compliance             | - |    |   | [-](https://github.com/open-telemetry/opentelemetry-python/issues/1109)    |    | -    |   |    |   | -  | -   |   X    |

Author: I'm not really sure what this spec is :) As no SDK has implemented it yet assuming it's optional.

|[Zipkin](specification/trace/sdk_exporters/zipkin.md)|
|Zipkin V1 JSON                                |   | +  |   | [-](https://github.com/open-telemetry/opentelemetry-python/issues/1173)    |    | -    | - | -  |   | -  | -   |   X    |
|Zipkin V1 Thrift                              |   | +  |   | [-](https://github.com/open-telemetry/opentelemetry-python/issues/1174)    |    | -    | - | -  |   | -  | -   |   X    |

> |Zipkin V2 JSON                                | + | +  |   | +    |    | -    | + | +  |   | +  | +   |        |

Reviewer 5: Are we sure we (can) support both V2 Zipkin formats before going Tracing Stable? I don't think most SIGs have cycles to implement **both** (we can ask/discuss this in the next Maintainers meeting).

Author: I thought I had left a comment but can't find it so maybe missed it. My intention with this matrix is long-term features. I figured language SIGs would determine how much of the table is implemented for a stable release. The optional I marked here are ones language SIGs could choose never to support. This is mostly legacy formats for exporters I think.

| OpenCensus                                            |          |    |                                                                       |    |                                                                         |      |        |     |      |     |      |       |
| TBD                                                   |          |    |                                                                       |    |                                                                         |      |        |     |      |     |      |       |
| Prometheus                                            |          |    |                                                                       |    |                                                                         |      |        |     |      |     |      |       |
| TBD                                                   |          |    |                                                                       |    |                                                                         |      |        |     |      |     |      |       |

> * For each type of exporter, OTLP, Zipkin, and Jaeger, implementing one of the supported formats is required,
>   while other formats can be considered optional.

Reviewer 6: ```suggestion
* For each type of exporter, OTLP, Zipkin, and Jaeger, implementing at least one of the supported formats is required.
Implementing more than one formats is optional.
```
